Automattic/simplenote-electron
# Fix: Simperium sync issues

**Status**

Ready to merge 🎉 

---

Closes #1579 

This PR changes the way that the editor's contents and Simperium interact; it's dependent on the changes made in `simperium@1.0.0`.

Previously we held content in the editor component and would delay sending those changes to the underlying Simperium client library (`node-simperium`) until the timeout flushed. This made for awkward interactions when we would received remote updates from the server that happened before that timeout flushed; this led to note corruption.

After this PR we are announcing immediately to the underlying client library when we make changes and telling it (via `{ sync: false }` in `noteBucket.update()`) that it's okay to wait to synchronize those changes with the server. We still maintain our debounce, but instead of sending the editor component's text on the timeout we more simply tell the underlying library to flush its changes (via `noteBucket.touch()`). The effect is that we've removed a source of concurrency bugs with those remote updates.

After this PR we should be removing a number of related data loss and note corruption issues all relating to the interplay of queued and received changes.

**Testing**

See detailed testing instructions in p2XJRt-1jm-p2.
Most testing needs are to open up builds of the app with these changes and to exercise as many update interactions as possible. In #1579, for example, we triggered these updates via the following test:

Open the app on two devices: call one `Local` and the other `Remote`

1. Create a note with contents `AC` and let it synchronize to both apps/devices
2. Disconnect `Local` from the network
3. On `Local` add a `D` to the end of the note to make `ACD`
4. With `Remote` still connected to the network, add a `B` in the middle to make `ABC`
5. After a few seconds (to ensure that `Remote` synchronizes with the server) re-connect `Local`
6. Verify the state of the note on both apps/devices

Before this change we were finding variations of `ACDD`, `ACD`, `AC`, and more incorrect end-states. After this change we should find what we expect: `ABCD`
